packages/drivers/odsp-driver/src/handleReadonlyServerState.ts

```diff
@@ -0,0 +1,56 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert, delay } from "@fluidframework/common-utils";
+import { getRetryDelayFromError } from "@fluidframework/driver-utils";
+import { OdspErrorType } from "@fluidframework/odsp-driver-definitions";
+import { FetchType } from "./epochTracker";
+
+export async function handleReadOnlyServerState<T>(
+    api: () => Promise<T>,
+    logger: ITelemetryLogger,
+    fetchType: FetchType,
+) {
+    let result: T | undefined;
+    let success = false;
+    let retryAfterMs = 1000; // has to be positive!
+    let numRetries = 0;
+    const maxRetries = 5;
+    let lastError: any;
+    do {
+        try {
+            result = await api();
+            success = true;
+        } catch (error) {
+            lastError = error;
+            // If the innerMostError is not "ServiceReadonly" error or we have retried the max times, then just
+            // give up.
+            if (error?.innerMostErrorCode !== OdspErrorType.serviceReadOnly || numRetries >= maxRetries) {
+                break;
+            }
+            numRetries++;
+            // If the error is about server becoming readonly for sometime, then wait for the specified time
+            // before retrying. If the waitTime is not specified, then we start with retrying immediately
+            // to max of 8s.
+            retryAfterMs = getRetryDelayFromError(error) ?? Math.min(retryAfterMs, 8000);
```

Comment:
>getRetryDelayFromError [](http://example.com/codeflow?start=27&length=22)

Does there happen to be "canRetry" in their response?
I'm curious if we need to bake the idea that this facet is retriable on our own and specifically, or if we should be respecting an existing and more generic marker.

This could be more clear if some sample response(s) for the sceanrio were included as a part of the review.

Reply:
Sorry I failed the mention the issue link in the description. You can take a look. 
The error that we are evaluating here is generated by driver code by looking at the server response. So this checks if that error has any wait time baked in the error. Their response for this specific error does not have any wait time.